### PR TITLE
Cache ordered partitions per event type

### PR DESCRIPTION
**Pull request by a1exsh**
Move sorting out of the hot path on batch publishing.

Based on findings from #1398

### CODE REVIEW

**Comment by a1exsh** [comment]
Note that this makes us sort partitions for _random_-partitioned event types as well, but also only once per publishing batch.

We can improve it further by refactoring the `PartitionStrategy` interface and the `PartitionResolver` component, but I don't see a clean way to do that for now.

File: core-services/src/main/java/org/zalando/nakadi/partitioning/PartitionResolver.java [hunk 1]
```diff
@@ -57,25 +57,18 @@ public void validate(final EventTypeBase eventType) throws NoSuchPartitionStrate
...
             throws PartitioningException {
-        final PartitionStrategy partitionStrategy = getPartitionStrategy(eventType);
-        final List<String> partitions = getPartitions(eventType);
 
-        return partitionStrategy.calculatePartition(nakadiRecordMetadata, partitions);
-    }
-
-    private List<String> getPartitions(final EventType eventType) {
-        return timelineService.getTopicRepository(eventType)
-                .listPartitionNames(timelineService.getActiveTimeline(eventType).getTopic());
```
[the remark this answers is not in the document]
Comment (a1exsh) [reply core-services/src/main/java/org/zalando/nakadi/partitioning/PartitionResolver.java:78]: Oops, I thought I did that :p

File: core-services/src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java [hunk 2]
```diff
@@ -251,22 +251,27 @@ public List<Timeline> getAllTimelinesOrdered(final String eventType)
...
             throw new TimelineException("Failed to get timeline", e);
         }
     }
 
+    public static Optional<Timeline> getActiveTimeline(final List<Timeline> timelines) {
+        final ListIterator<Timeline> rIterator = timelines.listIterator(timelines.size());
+        while (rIterator.hasPrevious()) {
+            final Timeline toCheck = rIterator.previous();
+            if (toCheck.getSwitchedAt() != null) {
+                return Optional.ofNullable(toCheck);
```
Comment (antban) [inline core-services/src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java:269]: `Optional.of` would be more suitable here

**Comment by antban** [comment]
:+1:

**Pull request merged**